lib/doctor.js

```diff
@@ -0,0 +1,107 @@
+var path = require('path')
+var chain = require('slide').chain
+var table = require('text-table')
+var color = require('ansicolors')
+var styles = require('ansistyles')
+var semver = require('semver')
+var which = require('which')
+var npm = require('./npm.js')
+var ping = require('./ping')
+var ansiTrim = require('./utils/ansi-trim')
+var output = require('./utils/output.js')
+var defaultRegistry = require('./config/defaults').defaults.registry
+var checksumCachedFiles = require('./doctor/checksum-cached-files')
+var checkFilesPermission = require('./doctor/check-files-permission')
+var getLatestNodejsVersion = require('./doctor/get-latest-nodejs-version')
+var getLatestNpmVersion = require('./doctor/get-latest-npm-version')
+var globalNodeModules = path.join(npm.config.globalPrefix, 'lib', 'node_modules')
+var localNodeModules = path.join(npm.config.localPrefix, 'lib', 'node_modules')
+
+module.exports = doctor
+
+doctor.usage = 'npm doctor'
+
+function doctor (args, silent, cb) {
+  args = args || {}
+  if (typeof cb !== 'function') {
+    cb = silent
+    silent = false
+  }
+
+  var actionsToRun = [
+    [ping, {}, true],
+    [getLatestNpmVersion],
+    [getLatestNodejsVersion, args['node-url']],
+    [which, 'git'],
+    [checkFilesPermission, npm.cache, 4],
```

Comment:
`filemode & 0400` would be true for files you have read access to. Should this be rewrite instead? (`6`)